### DISCUSSION TITLE
Allow Apps to request static --filesystem= permission

### DISCUSSION THREAD

**Discussion by boredsquirrel** [flatpak/xdg-desktop-portal#1405] [Ideas]
Hey, I would like to share, why I think that we will never have portals working for every app, and that this is normal and instead a different method should be used.

---

The portal method doesnt work with many apps.

I would expect an app to implement a check if some file exists, if it doesnt send some message to the system with an info "I want this permission", the system opens a popup "APP wants permanent permission to X", I click yes and then it gets that permission permanently. 

Nextcloud on MacOS does this for example. Kinda like on Android.

Instead when using portals, mounted directories have weird /var/run/ directories with strange IDs, and apps may look at the wrong place after a restart.

Instead, the portal could include a checkbox "permanent". If this is used, a flatpak override is done to add a permanent access to the real dir path, not mounted. A note "this is only applied on restart" would make sense.

This is especially needed for custom locations outside the xdg dirs, like /home/user/sync/pictures/random/edits for some image editing app that only looks in /home/user/pictures.

This system just doesnt work? Users should not need to enter the save directory every time they open an app.

With the current system, static filesystem permissions will always be needed, and that is fine, the integration of those should just be improved.

**Comment by rusty-snake** [comment]
> Allow Apps to request static --filesystem= permission

In order for a change in a static permission to take effect, a **app restart is required**.

This is a interruption in the workflow for the user and the app looses all runtime state.

**Reply by boredsquirrel** [reply]
this is a good point and indeed an issue, but could still work. Apps could set themselves up and then restart themselves, or is this not possible?

**Reply by rusty-snake** [reply]
We would need a restart portal.

**Comment by Mikenux** [comment]
Thanks for reporting this issue here.

Here are some questions to clarify things:

> I would expect an app to implement a check if some file exists, if it doesnt send some message to the system with an info "I want this permission", the system opens a popup "APP wants permanent permission to X", I click yes and then it gets that permission permanently.

I really don't understand. Could you tell what is the case by giving a concrete example?

> have weird /var/run/ directories with strange IDs

What do you expect instead? Filenames? Filepaths? And most importantly, for what purpose?

> lose the configuration after an app restart, ...

Could you tell what configuration? Remember the save directory?

> Users should not need to enter the save directory every time they open an app.

What are you expecting? Have access to a permanent save location like` ~/Downloads` for web browsers?

Thanks!

**Reply by Mikenux** [reply]
@boredsquirrel

**Reply by boredsquirrel** [reply]
1. The difference between a portal and a permanent filesystem permission to my knowledge is, that a portal mounts the dir to /run/ something and the filepaths are broken, the location may be broken even after app restart etc.

Many apps are simply broken with this.

If an app needs permanent access to a directory by default, that should be in the static filesystem permissions already present when installing.

I see how this may correlate with the portal option "give access to all files in this folder", but this is different.

I expect to grant apps access to a folder when needed, and that folder is available, not mounted or something. I currently do this in the KDE Settings and it is a pain tbh.

Example: KeepassXC should work without filesystem permissions, but it expects to save a location to a file to auto-open it on start. It uses the portal but is still broken unless I give it access to this directory.

This might be solved with a checkbox "grant permanent access" in the file chooser portal, doing an override. For the runtime, the app has access, and after restart it has permanent access. I dont know how well this would work, and KeepassXC would be broken once I guess, as the /run/... filepath is not the real one.

> What do you expect instead? Filenames? Filepaths? And most importantly, for what purpose?

I expect to grant apps access to the real filesystem location, not mount something somewhere in a cryptic location. I think a checkbox to do this would be good UX and I also think this is not always needed.

But keep in mind that 2 options are confusing for some people. But a "permanent" checkbox would make sense.

> Remember the save directory?

Sorry bad phrasing. Their config shows a filepath that doesnt exist anymore. Like in KeepassXC.

> Have access to a permanent save location like ~/Downloads for web browsers?

No, this is already possible. xdg-directories are not an issue here.

I want to open a random /home/user/sync/pictures/phone/edits folder with image-editor-123 and click "allow access permanently". 

As a restart is needed, I see how this is difficult. Either some form of stacking could work, to grant access to multiple things, then restart. Or simply a note "this access is only there after a restart", maybe with a button in the filechooser portal?

**Reply by Mikenux** [reply]
Logically, an app can restore access to any file it has access to, at least when that file was opened or saved using the file selection portal previously (using the document store). Maybe the KeepPassXC developers don't know how to restore the access? @hfiguiere

> Sorry bad phrasing. Their config shows a filepath that doesnt exist anymore. Like in KeepassXC.

If you are talking about showing the real filepath in the app UI, this feature has been added, but it is for the next stable release of xdg-desktop-portal.

**Reply by boredsquirrel** [reply]
So:

1. Apps can store filepaths in the "document store" which may be different from what they need as native app
2. The portal will give the apps access to the real file path name, even though it may use a different one?

This sounds like it could be a workaround or solution for this issue.

**Reply by Mikenux** [reply]
Here is if you want to look at the code:
- https://github.com/flatpak/xdg-desktop-portal/pull/1364
- https://github.com/flatpak/xdg-desktop-portal/pull/1372

> what they need as native app

What does a native app need? See the filesystem as native (like when directories are mounted using static permissions; e.g.` /home/user/pictures`)? If so, what are all the benefits for developers over using/manipulating the document store?

**Reply by Mikenux** [reply]
Also, let's say that having permanent access to files for an app is no longer possible (I don't say it will be the case):

> Example: KeepassXC should work without filesystem permissions, but it expects to save a location to a file to auto-open it on start. It uses the portal but is still broken unless I give it access to this directory.

You are talking about a file, which does not need access to a whole directory. Which kind of file it is? (file needing permanent write access? if so, what it's purpose is?)

> I want to open a random /home/user/sync/pictures/phone/edits folder with image-editor-123 and click "allow access permanently".

Why needing permanent access to a folder if it is to just open a file? Isn't using the file manager or file chooser each time sufficient?

**Reply by boredsquirrel** [reply]
a native app is an app that runs unsandboxed and sees the entire filesystem

keepassxc needs access to the file that it opened previously. a .kdbx file, a keepass database.

This file is opened with the portal, but then an incorrect path is shown. Instead, when giving the app access to the exact location of that file manually and then opening the file, afaik the correct path this shown and it can auto-open it on next restarts.

no, the point is that the filemanager is not a good system for every app. Imagine a gallery app, this is the filemanager, but should have access to multiple directories, also custom ones. This is not easy to do, not possible on GNOME without flatseal.

I open my password manager, that unlocks the vault. I dont open the kdbx file. 

The filemanager method works often but also often not. A ton of apps have internal filemanagers, which you may not use (like in libreoffice) or may use.

**Reply by Erick555** [reply]
> keepassxc needs access to the file that it opened previously. a .kdbx file, a keepass database.

portals support this

> This file is opened with the portal, but then an incorrect path is shown

what do you mean by incorrect path? Even if the path looks confusing for human it's not a problem for app.

libreofice doesn't support portals

**Reply by Mikenux** [reply]
@boredsquirrel: I installed KeePassXC (version 2.7.9 from flathub) and removed the `filesystem=host` permission. I tested it and it works. When I launch KeePassXC, I type the password I set to unlock and it works. Only the host path is not displayed, but, as I said, this was only implemented recently.

My versions:
- Fedora Linux Workstation (GNOME 46).
- flatpak: 1.15.10.
- xdg-desktop-portal: 1.18.4.

**Reply by boredsquirrel** [reply]
@Mikenux so if you close it again and reopen it, can it automatically open the previously used DB without needing to select it again from a filemanager?

Then this may not be needed in this case

**Reply by Mikenux** [reply]
Yes.

**Comment by Mikenux** [comment]
Maybe for libraries/collections (music, pictures, games) if performance is an issue or there are other issues that are not fixable in a reasonable time (regarding the FUSE filesystem).

**Reply by boredsquirrel** [reply]
not sure what fuse has to do here.

Afaik this is needed for many apps that are kind of filemanagers themselves. that display the name of a file and its correct location.

These locations would be incorrect if they only used portals.

And not having access to the correct files and directories may cause issues when passing these files on to other apps? not sure.

**Reply by Erick555** [reply]
portal use fuse filesystem

**Reply by Mikenux** [reply]
> And not having access to the correct files and directories may cause issues when passing these files on to other apps? not sure.

This is problematic if the user uses it to copy and paste. But developers use the `host` or `home` filesystem permission to show the correct path.

**Comment by boredsquirrel** [comment]
Snap just got a portal update that at least added the ability to grant access to a folder.

But snap is also different, using apparmor, and maybe better for displaying the actual filepath and not a mounted one

**Reply by Erick555** [reply]
snap uses same portal as flatpak. folder support exist for along time.

**Reply by boredsquirrel** [reply]
may only be implemented in GNOME then

**Reply by Mikenux** [reply]
Are you using flatpak or snap? What are your versions (flatpak/snap, xdg-desktop-portal, desktop environment)?
